translate/storage/markdown.py

```diff
@@ -0,0 +1,462 @@
+
+            # render the translatable content (with placeholders) to markdown
+            content_md = "\n".join(
+                self.fragments_to_lines(content, max_line_length=float("inf"))
+            )
+
+            # translate and parse into new fragments. handle hard line breaks.
+            if content_md:
+                translated_md = self.translate_callback(content_md, self.path)
+                translated_md = translated_md.replace("\n", "\\\n").strip(" \t")
+                translated_md = self.remove_placeholder_markers(
+                    translated_md, placeholders
+                )
+                translated = self.make_fragments(
+                    span_token.tokenize_inner(translated_md)
+                )
+                fragments = chain(leader, translated, trailer)
+
+            # expand placeholders and render into final markdown.
+            # list-ify to let all generators run before exiting the try/finally block
+            expanded = list(self.expand_placeholders(fragments))
+            return super().fragments_to_lines(expanded, max_line_length=max_line_length)
+        finally:
+            self.bypass = False
+
+    @classmethod
+    def merge_adjacent_placeholders(
+        cls, fragments: Iterable[Fragment]
+    ) -> Iterable[Fragment]:
+        """Replaces sequences of placeholders and whitespace with larger placeholders."""
+        fragments = list(fragments)
+        start = 0
+        while start < len(fragments):
+            if getattr(fragments[start], "placeholder_content", None):
+                end = 0
+                for j in range(start + 1, len(fragments)):
+                    if getattr(fragments[j], "placeholder_content", None):
+                        end = j + 1
+                    elif fragments[j].text.isspace():
+                        pass
+                    else:
+                        break
+                if end:
+                    chunk = fragments[start:end]
+                    placeholder = Fragment(None, placeholder_content=chunk)
+                    placeholder.important = any(
+                        getattr(fragment, "important", False) for fragment in chunk
+                    )
+                    fragments[start:end] = (placeholder,)
+            start += 1
+
+        return fragments
+
+    @classmethod
+    def trim_flanking_placeholders(
+        cls, fragments: Iterable[Fragment]
+    ) -> Tuple[Iterable[Fragment], Iterable[Fragment], Iterable[Fragment]]:
+        """
+        Splits leading and trailing placeholders and whitespace, and the main
+        content, into separate lists. Placeholders marked as important are kept
+        with the main content.
+        """
+        fragments = list(fragments)
+
+        l = 0
+        while l < len(fragments):
+            if getattr(fragments[l], "placeholder_content", None):
+                if getattr(fragments[l], "important", False):
+                    break
```

Comment:
Can we have tests for this, please?

Reply:
This function is tested by the test cases `test_remove_placeholders_from_both_ends_of_translation_units`, `test_paragraph_with_only_whitespace_and_placeholders`, `test_plain_image`, and `test_plain_link`.

Would you still like some more tests targeting this particular function directly?

Reply:
The break statement is never reached while running tests. So it seems that important placeholders are not tested.

Reply:
Right. This is because the only placeholders considered "important" at the moment are links, and they are always on the trailing side. I'd rather not break the leading/trailing symmetry, though, so I'll add a unit test for this.